Tweak the rows that get returned by the pivot QP

camsaul opened the pull request:
I wrote a bunch of tests and simplified things a bit as I was trying to figure out what the code was supposed to do and how it worked

Fixes #14329

camsaul commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
+
+    (breakout-combinations 3 [1 2] nil) ;; -> [[0 1 2] [] [1 2] [2] [1]]"
+  [num-breakouts pivot-rows pivot-cols]
+  ;; validate pivot-rows/pivot-cols
+  (doseq [[k pivots] {:pivot-rows pivot-rows, :pivot-cols pivot-cols}
+          i          pivots]
+    (when (>= i num-breakouts)
+      (throw (ex-info (tru "Invalid {0}: specified breakout at index {1}, but we only have {2} breakouts"
+                           (name k) i num-breakouts)
+                      {:type          qp.error-type/invalid-query
+                       :num-breakouts num-breakouts
+                       :pivot-rows    pivot-rows
+                       :pivot-cols    pivot-cols}))))
+  (sort-by
+   (partial group-number num-breakouts)
+   (distinct
+    (map
+     vec
+     (concat
+      ;; e.g. given num-breakouts = 4; pivot-rows = [0 1 2]; pivot-cols = [3]
+      ;; subtotal rows
+      ;; _.range(1, pivotRows.length).map(i => [...pivotRow.slice(0, i), ...pivotCols])
+      ;;  => [0 3] [0 1 3] => 1001 1101
+      (for [i (range 1 (count pivot-rows))]
+        (concat (take i pivot-rows) pivot-cols))
+      ;; “row totals” on the right
+      ;; pivotRows
+      ;; => [0 1 2] => 1110
+      [pivot-rows]
+      ;; subtotal rows within “row totals”
+      ;; _.range(1, pivotRows.length).map(i => pivotRow.slice(0, i))
+      ;; => [0] [0 1] => 1000 1100
+      (for [i (range 1 (count pivot-rows))]
+        (take i pivot-rows))
+      ;; “grand totals” row
+      ;; pivotCols
+      ;; => [3] => 0001
+      [pivot-cols]
+      ;; bottom right corner [] => 0000
+      [[]])))))
```
I separated out the code that determines the combination of breakout indecies to generate queries for with the code that actually puts together the corresponding breakouts to make it easier to test and make determining the group number of the breakouts combination easier (no need to reverse the breakout back to its index later on)

camsaul commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
-                             #(bit-or (bit-shift-left 1 %2) %1) 0
-                             (map #(get bitmask-index %) breakout))]
-             (add-grouping-field request breakout
-                                 (bitwise-not-truncate (count all-breakouts) start-mask))))
-         new-breakouts)))
+  [{{all-breakouts :breakout} :query, :keys [pivot-rows pivot-cols query], :as outer-query}]
+  (try
+    (for [breakout-indecies (breakout-combinations (count all-breakouts) pivot-rows pivot-cols)
+          :let              [group-number    (group-number (count all-breakouts) breakout-indecies)
+                             new-breakouts (for [i breakout-indecies]
+                                             (nth all-breakouts i))]]
+      (add-grouping-field outer-query new-breakouts group-number))
+    (catch Throwable e
+      (throw (ex-info (tru "Error generating pivot queries")
+                      {:type qp.error-type/qp, :query query}
+                      e)))))
```
simplified a bit

camsaul commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
+                      {:type qp.error-type/qp, :query query}
+                      e)))))
+
+;; this function needs to be executed at the start of every new query to
+;; determine the mapping for maintaining query shape
+(defn- map-cols
+  {:arglists '([query context])}
+  [query {all-expected-cols ::all-expected-columns}]
+  (assert (seq all-expected-cols))
+  (let [query-cols (map-indexed vector (qp/query->expected-cols query))]
+    (map (fn [item]
+           (some #(when (= (:name item) (:name (second %)))
+                    (first %))
+                 query-cols))
+         all-expected-cols)))
+
+;; this function needs to be called for each row so that it can actually
+;; shape the row according to the `map-cols` fn above
+(defn- map-row
+  [row cols]
+  ;; the first query doesn't need any special mapping, it already has all the columns
+  (if cols
+    (for [i cols]
+      (when i
+        (nth row i)))
+    row))
```
I moved this stuff out of `run-pivot-query` so it could be closer to where it's actually used

camsaul commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
+  "Come up with a display name given a combination of breakout `indecies` e.g.
 
-(defn- generate-specified-breakouts
-  "Generate breakouts specified by pivot_rows"
```
Queries get normalized to lisp-case so I suspect trying to use `snake_case` keys like `pivot_rows` and `pivot_cols` is going to cause problems, so I changed it. We should only use `snake_case` keys for things that directly touch the DB/API boundary e.g. `defendpoint` forms or DB code. Everything else should use `lisp-case`... convert keys in the API endpoints if you need to. Once I get #14113 working this won't be a thing anymore

camsaul commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -119,31 +165,20 @@
-                                        col-mapping)
-                                   row))))))))
+   ;; make sure the query keys are normalized to lisp-case.
+   (let [query (normalize/normalize query)]
+     (qp.store/with-store
```
code will still work with `snake_case` keys (not sure if this actually is needed)

camsaul commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
+  ;; validate pivot-rows/pivot-cols
+  (doseq [[k pivots] {:pivot-rows pivot-rows, :pivot-cols pivot-cols}
+          i          pivots]
+    (when (>= i num-breakouts)
+      (throw (ex-info (tru "Invalid {0}: specified breakout at index {1}, but we only have {2} breakouts"
+                           (name k) i num-breakouts)
+                      {:type          qp.error-type/invalid-query
+                       :num-breakouts num-breakouts
+                       :pivot-rows    pivot-rows
+                       :pivot-cols    pivot-cols}))))
```
added some extra validation

camsaul commented on test/metabase/query_processor/pivot_test.clj:
```diff
@@ -15,54 +16,167 @@
+                 (map :display_name (mt/cols results))))
+        (is (apply distinct? rows))
+        ;; OLD RESULTS are commented out below
+        (is (= [;; ["Doohickey" "Facebook" "2019-01-01T00:00:00Z" 0  263 ]
+                ;; ["Doohickey" "Facebook" "2020-01-01T00:00:00Z" 0  89  ]
+                ;; ["Doohickey" "Google"   "2019-01-01T00:00:00Z" 0  276 ]
+                ;; ["Doohickey" "Google"   "2020-01-01T00:00:00Z" 0  100 ]
+                ;; ["Gizmo"     "Facebook" "2019-01-01T00:00:00Z" 0  361 ]
+                ;; ["Gizmo"     "Facebook" "2020-01-01T00:00:00Z" 0  113 ]
+                ;; ["Gizmo"     "Google"   "2019-01-01T00:00:00Z" 0  325 ]
+                ;; ["Gizmo"     "Google"   "2020-01-01T00:00:00Z" 0  101 ]
+                ;; [nil         nil        nil                    7  1628]
+                ["Doohickey" "Facebook" "2019-01-01T00:00:00Z" 0  263 ]
+                ["Doohickey" "Facebook" "2020-01-01T00:00:00Z" 0  89  ]
+                ["Doohickey" "Google"   "2019-01-01T00:00:00Z" 0  276 ]
+                ["Doohickey" "Google"   "2020-01-01T00:00:00Z" 0  100 ]
+                ["Gizmo"     "Facebook" "2019-01-01T00:00:00Z" 0  361 ]
+                ["Gizmo"     "Facebook" "2020-01-01T00:00:00Z" 0  113 ]
+                ["Gizmo"     "Google"   "2019-01-01T00:00:00Z" 0  325 ]
+                ["Gizmo"     "Google"   "2020-01-01T00:00:00Z" 0  101 ]
+                ;; [nil         "Facebook" "2019-01-01T00:00:00Z" 1  624 ]
+                ;; [nil         "Facebook" "2020-01-01T00:00:00Z" 1  202 ]
+                ;; [nil         "Google"   "2019-01-01T00:00:00Z" 1  601 ]
+                ;; [nil         "Google"   "2020-01-01T00:00:00Z" 1  201 ]
+                ;; ["Doohickey" nil        "2019-01-01T00:00:00Z" 2  539 ]
+                ;; ["Doohickey" nil        "2020-01-01T00:00:00Z" 2  189 ]
+                ;; ["Gizmo"     nil        "2019-01-01T00:00:00Z" 2  686 ]
+                ;; ["Gizmo"     nil        "2020-01-01T00:00:00Z" 2  214 ]
+                ;; [nil         nil        "2019-01-01T00:00:00Z" 3  1225]
+                ;; [nil         nil        "2020-01-01T00:00:00Z" 3  403 ]
+                ["Doohickey" "Facebook" nil                    4  352 ]
+                ["Doohickey" "Google"   nil                    4  376 ]
+                ["Gizmo"     "Facebook" nil                    4  474 ]
+                ["Gizmo"     "Google"   nil                    4  426 ]
+                ;; [nil         "Facebook" nil                    5  826 ]
+                ;; [nil         "Google"   nil                    5  802 ]
+                ["Doohickey" nil        nil                    6  728 ]
+                ["Gizmo"     nil        nil                    6  900 ]
+                [nil         nil        nil                    7  1628]]
+               rows))))))
```
here's the results that were getting returned (I copied @paulrosenzweig 's example in #14329 directly) with the duplicate/unneeded rows commented out once I fixed the behavior

camsaul commented on test/metabase/query_processor/pivot_test.clj:
```diff
@@ -15,54 +16,167 @@
+      (let [results (pivot/run-pivot-query (test-query))
+            rows    (mt/rows results)]
+        ;; TODO - fix me
+        #_(is (= ["Category"
+                  "Source"
+                  "Created At"
+                  "pivot-grouping"
+                  "Count"]
+                 (map :display_name (mt/cols results))))
```
the columns that come back in metadata for the `run-pivot-query` function are wrong... not sure if there's an open issue to fix this or not yet

robdaemon replied:
Ahhh that's likely an issue from the reordering of the columns that occurred with fixing the previous pgsql issue. There is no open issue for it.

camsaul replied:
fixed it

camsaul commented on test/metabase/query_processor/pivot_test.clj:
```diff
@@ -15,54 +16,167 @@
+         clojure.lang.ExceptionInfo
+         #"Invalid pivot-cols: specified breakout at index 3, but we only have 3 breakouts"
+         (#'pivot/breakout-combinations 3 [] [0 1 2 3]))))
+  ;; TODO -- we should require these columns to be distinct as well (I think?)
+  ;; TODO -- require all numbers to be positive
+  ;; TODO -- can you specify something in both pivot-rows and pivot-cols?
```
there's some more stuff we should add validation for. For example we shouldn't let you use negative indices in `pivot-rows` or include an index in both

camsaul commented on test/metabase/query_processor/pivot_test.clj:
```diff
@@ -15,54 +16,167 @@
+             [     ]])
+           (#'pivot/breakout-combinations 4 [0 1 2] [3])))))
+
+(deftest group-number-test
+  (doseq [[indecies expected] {[0]     6
+                               [0 1]   4
+                               [0 1 2] 0
+                               []      7}]
+    (is (= expected
+           (#'pivot/group-number 3 indecies)))))
```
I'm not sure exactly why the group numbers are determined this way but I made sure to match existing behavior. I was considering changing to group *names* like `001` or `010` but didn't want to break the FE

robdaemon replied:
That would break the frontend. See above comment about the bitmask calculation.

robdaemon commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
-                      ;;TODO: replace this value with a bitmask or something to indicate the source better
                       (update-in [:query :expressions]
-                                 #(assoc % "pivot-grouping" [:abs bitmask])))]
+                                 #(assoc % "pivot-grouping" [:abs group-number])))]
```
This is a bitmask, per https://www.postgresql.org/docs/current/queries-table-expressions.html#QUERIES-GROUPING-SETS at Paul's request, hence the name `bitmask`

camsaul replied:
gotcha... I renamed it back to `group-bitmask`

robdaemon commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
-         new-breakouts)))
+  [{{all-breakouts :breakout} :query, :keys [pivot-rows pivot-cols query], :as outer-query}]
+  (try
+    (for [breakout-indecies (breakout-combinations (count all-breakouts) pivot-rows pivot-cols)
```
s/indecies/indices/

robdaemon commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
-          :when (not (zero? (bit-and (bit-shift-left 1 item) combo)))]
-      (nth xs item))))
+(defn- group-number
+  "Come up with a display name given a combination of breakout `indecies` e.g.
```
indices

robdaemon commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
-                              (vec (range 0 (count breakouts)))))))))
+    (group-number 3 [1])   ; -> 5
+    (group-number 3 [1 2]) ; -> 1"
+  [num-breakouts indecies]
```
indices

robdaemon commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
+   (map (partial bit-shift-left 1))
+   (completing bit-xor)
+   (int (dec (Math/pow 2 num-breakouts)))
+   indecies))
```
indices

camsaul replied:
wow I really can't spell

robdaemon replied:
I can only spell because the red squigglies underneath words nowadays 😁

robdaemon commented on test/metabase/query_processor/pivot_test.clj:
```diff
@@ -15,54 +16,167 @@
+                ;; ["Gizmo"     "Google"   "2020-01-01T00:00:00Z" 0  101 ]
+                ;; [nil         nil        nil                    7  1628]
+                ["Doohickey" "Facebook" "2019-01-01T00:00:00Z" 0  263 ]
+                ["Doohickey" "Facebook" "2020-01-01T00:00:00Z" 0  89  ]
```
I thought it looked weird that we were considering these duplicates? Unless I was misunderstanding.

camsaul replied:
the rows that are commented out are the duplicates -- if you look at them they appear twice

paulrosenzweig commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -1,98 +1,144 @@
-                (seq pivot-rows))
-       (generator [(concat [(first pivot-rows)] pivot-cols)]))
-     (generator (powerset (or pivot-rows
-                              ;; this can happen for the public/embed endpoints,
-                              ;; where we aren't given a pivot_rows / pivot_cols
-                              ;; parameter, so we'll just generate everything
```
I forgot to mention in the issue what this comment calls out. We need to handle missing `pivot-rows` and `pivot-cols`.

The frontend doesn't have the query when public/embedded so it can't generate the list of indexes. Two possible hacks around that:
1. When those are missing/empty return a full powerset.
2. Duplicate the logic to generate rows/cols indices from query+vis settings in the backend.

This issue should go away when we build `breakout_sets` into MBQL. Then, the query itself will contain the needed information.

For now, either option works for me, but I prefer 1. It's slower for public/embedded since it runs unnecessary queries, but it avoids potentially inconsistent FE/BE logic duplication.

camsaul replied:
So... use the full powerset if cols/rows are missing. Otherwise use the rules from your issue

paulrosenzweig replied:
Yup that's right!

paulrosenzweig reviewed (commented):
I'll test it out in a bit, but looks good!

paulrosenzweig commented:
I'm seeing an issue where moving columns between pivot_cols and pivot_rows returns a different number of columns in the response. 

Example broken request:
```
{
  "type":"query",
  "query":{
    "source-table":2,
    "aggregation":[["count"]],
    "breakout":[
      ["fk->",["field-id",13],["field-id",4]],
      ["fk->",["field-id",11],["field-id",26]],
      ["datetime-field",["field-id",12],"year"],
      ["binning-strategy",["field-id",10],"default"]]
    },
  "database":1,
  "parameters":[],
  "pivot_rows":[0,1],
  "pivot_cols":[2,3]
}
```

The response is missing one column. It returns all columns except one for breakout[1] (`["fk->",["field-id",11],["field-id",26]]`).

Updating that request to have `"pivot_rows":[0,1,2,3],"pivot_cols":[]` correctly returns all six expected columns.

camsaul commented:
@paulrosenzweig can you check that the results make sense now? @robdaemon please re-review I reworked some more stuff and added a lot more tests to make sure everything is working correctly.

camsaul commented:
I also fixed the incorrect columns issues

camsaul commented on test/metabase/api/pivots.clj:
```diff
@@ -12,38 +13,41 @@
-         :breakout    [[:fk-> $orders.user_id $people.state]
-                       [:fk-> $orders.user_id $people.source]
-                       [:fk-> $orders.product_id $products.category]]})
-      (assoc :pivot_rows [1 0]
-             :pivot_cols [2])))
+  (mt/dataset sample-dataset
+    (-> (mt/mbql-query orders
+          {:aggregation [[:count] [:sum $orders.quantity]]
+           :breakout    [$orders.user_id->people.state
+                         $orders.user_id->people.source
+                         $orders.product_id->products.category]})
+        (assoc :pivot_rows [1 0]
+               :pivot_cols [2]))))
```
I reworked a bunch of these tests to use the more conscise `$ids` shorthand for `:fk->` forms

robdaemon replied:
I can't help but think we have too many different ways to do this.

camsaul replied:
it's all really just shorthand for the same thing... it makes it a lot more convenient to write queries from the REPL IMO or to write tests. I would rather write `$user_id->people.state` than `[:fk-> [:field-id (mt/id :orders :user_id)] [:field-id (mt/id :people :state)]]`. I think it's more readable this way too.

Added benefit: if we implement something like #14057 and replace `:field-id` and `:fk->` with a new unified `:field` clause or something like that we can tweak these macros to expand to the new replacement forms.

robdaemon commented on test/metabase/api/pivots.clj:
```diff
@@ -1,4 +1,6 @@
 (ns metabase.api.pivots
+  ;; TODO -- I think we should roll this all in to `metabase.query-processor.pivot-test` or maybe a new namespace
+  ;; called `metabase.query-processor.pivot.test-util`
```
yeah this is basically a `test-util` namespace, I wasn't sure quite how to name it

robdaemon commented on test/metabase/api/card_test.clj:
```diff
@@ -1453,19 +1453,30 @@
                   :collections       s/Any}
                  (mt/user-http-request :crowberto :get 200 (format "card/%s/related" (u/get-id card)))))))
 
+(defn assert-pivot-card-results-are-correct [result]
```
I avoided this because it will cause any assertion failure's message to show in this file, as opposed to where it was called in a test

example, when `(mt/rows result)` fails, the error you get is:

`Query does not have any :rows in results (query_processor_test.clj:316)`

instead of the test it fails in.

camsaul replied:
Good point. Maybe we should change `(mt/rows)` to do a clojure.test `is` assertion rather than throwing an Exception. It seems like that would make it more useful everywhere

camsaul committed:
Update tests

camsaul committed:
Simpler diff

camsaul committed:
Even smaller diff

camsaul committed:
Test fix :wrench:

camsaul committed:
Test fix :wrench:

robdaemon reviewed (approved):
merge away!

camsaul committed:
Remove unused fn :wrench:

camsaul committed:
Merge branch 'master' into fix-14329

camsaul committed:
Make CodeCov happy
